Control Non-Public Repository Processing

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR adds controls for two things:
- Whether or not to process non-public repositories (default: public only).
- If processing non-public repositories, whether or not to output those repository names in logging (default: mask repository names).

It also adds information about the Action's inputs/outputs to the README as that information was missing.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

We should neither process private repositories nor leak information about private repositories unless it is specifically enabled. The README should contain information about the Action's inputs/outputs for convenience and good documenting.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

<!-- How did you test your changes? How could someone else test this PR? -->
Automated tests pass. I tested that the two controls work as expected with manual `lineage_scan` runs.
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [x] All new and existing tests pass.

mcdonnnj committed:
Add controls for non-public repositories

Add functionality to only process public repositories. Add the ability to mask
non-public repo names in the logs to prevent data leakage.

mcdonnnj committed:
Add new inputs to Action metadata

Add the `mask_non_public_repos` and `public_repos_only` inputs to control that
functionality in the Python code.

mcdonnnj committed:
Add Action metadata information to README

Add information about the inputs and outputs for this Action to the README.

mcdonnnj committed:
Update logging configuration

Use a library defined value to set the log level. Set the output stream to
`sys.stdout` instead of `sys.stderr` so that GitHub Actions commands can
manipulate the output (secret masking, log grouping, etc).

dav3r commented on README.md:
```diff
@@ -38,7 +38,27 @@ lineage:
+| access_token | GitHub personal access token. | n/a | yes |
+| mask_non_public_repos | Control the visibility of private repo names in the logs. | `true` | no |
+| public_repos_only | Control non-public repository processing. | `true` | no |
+| repo_query | GitHub search query to find repositories for pull requests. | n/a | yes |
```
We should include a `Type` column in this table.  It may also be helpful to add example values for the required inputs (in the description field).

mcdonnnj replied:
All GitHub Action inputs are strings (because they are loaded into the environment as environment values) so I'm not sure how well a `Type` column would work as far as setting expectations. Would something more like `Interpreted Type` or `Converted Type` make more sense just to give a better idea of how the data is interpreted?

dav3r replied:
Good thinking- I like `Interpreted Type` for this.

mcdonnnj replied:
Please see cf52622 and 37a4531 and let me know if there are any adjustments you'd like to see made.

dav3r replied:
Those changes looks great to me - thanks!

dav3r reviewed (approved):
Thanks for locking this down. 🔐 
Please take a look at my one small request and comment.

mcdonnnj committed:
Add type information for Action metadata to the README

Add information about how inputs are interpreted to the README. Add a default
placeholder to the example table for outputs to describe the type of an output.

In both cases the value will be converted to a string because of how inputs and
outputs are processed. This information will just help set expectations for how
an input is interpreted and what kind of data an output represents.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Add more information about inputs to the README

Expand the description of non-basic inputs to better describe what they should
contain.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Standardize table formatting in the README

Makes the formatting for the three tables in the README consistent.

jsf9k commented on README.md:
```diff
@@ -38,7 +38,27 @@ lineage:
+| Name | Description | Interpreted Type | Default | Required |
+|------|-------------|------------------|---------|:--------:|
+| access_token | GitHub personal access token (see [GitHub's documentation](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token)). | `string` | n/a | yes |
+| mask_non_public_repos | Control the visibility of private repo names in the logs. | [`boolean`](https://yaml.org/spec/1.2.2/#1032-tag-resolution) | `true` | no |
+| public_repos_only | Control non-public repository processing. | [`boolean`](https://yaml.org/spec/1.2.2/#1032-tag-resolution) | `true` | no |
+| repo_query | GitHub search query to find repositories for pull requests (e.g. "org:cisagov archived:false"). | `string` | n/a | yes |
```
* I think "private" is a much more succinct way of saying "non-public".
* Your descriptions for the boolean values tell me that they "control" something but don't tell me when i should use `true` and when i should use `false`.
* The description for `repo_query` was a little awkward and unclear, so I touched it up.
```suggestion
| mask_private_repos | Whether to mask the names of private repositories in the logs. | [`boolean`](https://yaml.org/spec/1.2.2/#1032-tag-resolution) | `true` | no |
| include_private_repos | Whether to process private repositories. | [`boolean`](https://yaml.org/spec/1.2.2/#1032-tag-resolution) | `false` | no |
| repo_query | GitHub search query to use when finding repositories for which to create pull requests (e.g. "org:cisagov archived:false"). | `string` | n/a | yes |
```

mcdonnnj replied:
I only used "non-public" because that was a blanket term to cover both `private` and `internal` repositories. I'm not sure if the API exposes `internal` in repository information yet, but I thought I'd just future-proof the language.

jsf9k replied:
OK, I wasn't thinking of the "internal" repos.  What about the two other bullet points?

mcdonnnj replied:
Ah right, those descriptions were a bit rough so I'm happy to improve them.

jsf9k replied:
I also suggest renaming `public_repos_only` to `include_private_repos` (or similar) so that the variable name more clearly lets the user know what the variable controls.

mcdonnnj replied:
Would you take a look at 4afa7c4 and let me know what you think?

jsf9k commented on action.yml:
```diff
@@ -10,6 +10,14 @@ inputs:
   access_token:
     description: "GitHub personal access token."
     required: true
+  mask_non_public_repos:
+    default: true
+    description: "Control the visibility of private repo names in the logs."
+    required: false
+  public_repos_only:
+    default: true
+    description: "Control non-public repository processing."
+    required: false
   repo_query:
     description: "GitHub search query to find repositories for pull requests."
     required: true
```
Related to earlier suggestions:
```suggestion
  mask_private_repos:
    default: true
    description: "Whether to mask the names of private repositories in the logs."
    required: false
  include_private_repos:
    default: false
    description: "Whether to process private repositories."
    required: false
  repo_query:
    description: "GitHub search query to use when finding repositories for which to create pull requests."
    required: true
```

jsf9k commented on src/lineage/entrypoint.py:
```diff
@@ -269,14 +269,19 @@ def get_code_owners(repo: Repository.Repository) -> Generator[str, None, None]:
     access_token: Optional[str] = core.get_input("access_token")
     github_actor: Optional[str] = os.environ.get("GITHUB_ACTOR")
     github_workspace_dir: Optional[str] = os.environ.get("GITHUB_WORKSPACE")
+    mask_non_public: bool = core.get_boolean_input("mask_non_public_repos")
     repo_query: Optional[str] = core.get_input("repo_query")
+    public_only: bool = core.get_boolean_input("public_repos_only")
```
```suggestion
    mask_private: bool = core.get_boolean_input("mask_private_repos")
    repo_query: Optional[str] = core.get_input("repo_query")
    include_private: bool = core.get_boolean_input("include_private_repos")
```

jsf9k commented on src/lineage/entrypoint.py:
```diff
@@ -318,6 +323,13 @@ def main() -> None:
     for repo in repos:
+        # Extra controls if the repo is private
+        if repo.private:
+            if public_only:
+                continue
+            # Ensure that non-public repo names do not show up in the logs
+            if mask_non_public:
```
```suggestion
            if not include_private:
                continue
            # Ensure that non-public repo names do not show up in the logs
            if mask_private:
```

jsf9k reviewed (commented):
I suggest renaming a few input variables and tweaking some variable descriptions.  Let me know what you think.

mcdonnnj committed:
Improve input name and descriptions

Improve the descriptions for a few inputs and the name for one to make it
easier to discern the functionality of these inputs.

Co-authored-by: Shane Frasier <jeremy.frasier@trio.dhs.gov>

jsf9k reviewed (approved):
Thanks for taking my comments into consideration!

dav3r reviewed (approved):
I'm good with how this turned out. 👍
